sockeye/utils.py

```diff
@@ -264,6 +264,124 @@ def average_tensors(tensors: List[pt.Tensor]) -> pt.Tensor:
+        :return (adjusted) first_steps_masking (batch_size * beam_size, max target prefix length + 1).
+
+        An illustrative example of how first_step_masking is adjusted
+
+        Inputs:
+
+        target_prefix (batch_size = 2, max target prefix length = 2)
+
+        tensor([1 2]
+               [1 0])
+        Note: Two target prefix tokens in the first sentence, \
+        one target prefix token in the second sentence.
+
+        first_step_mask (batch_size = 2 * beam_size = 5, 1)
+
+        tensor([[0],
+        [inf],
+        [inf],
+        [inf],
+        [inf],
+        [0],
+        [inf],
+        [inf],
+        [inf],
+        [inf])
+
+        Output:
+        Adjusted first_step_mask (batch_size * beam_size, max target prefix length + 1):
+
+        tensor([[0 0 0],
+                [inf inf inf],
+                [inf inf inf],
+                [inf inf inf],
+                [inf inf, inf],
+                [0 0 0],
+                [inf inf 0],
+                [inf inf 0],
+                [inf inf 0],
+                [inf inf 0]])
+
+        The concrete steps of what this function does are as follows:
+
+        Step 1: Create a zero masking matrix with shape (batch size, max target prefix length + 1)
+        Fill 1 into this masking matrix based on the target prefix
+
+        target prefix  initialize masking    masking      roll one step to the right
+                       from target prefix    is not 0      and assign 1 at index 0
+            [1 2]    ->    [1 2 0]        -> [1 1 0]  ->           [1 1 1]
+            [1 0]          [1 0 0]           [1 0 0]               [1 1 0]
+
+        Step 2: Adjust first_step_mask based on masking
+
+        masking     expand masking with     expand first_step_mask with max target
+                         beam size         prefix length, fill 0 where masking is 0
+        [1 1 1]  ->      [1 1 1]        ->             [0 0 0]
+        [1 1 0]          [1 1 1]                       [inf inf inf]
+                         [1 1 1]                       [inf inf inf]
+                         [1 1 1]                       [inf inf inf]
+                         [1 1 1]                       [inf inf inf]
+                         [1 1 0]                       [0 0 0]
+                         [1 1 0]                       [inf inf 0]
+                         [1 1 0]                       [inf inf 0]
+                         [1 1 0]                       [inf inf 0]
+                         [1 1 0]                       [inf inf 0]
+        """
+        first_step_mask_sizes = first_step_mask.size()
+        target_prefix_sizes = target_prefix.size()
+        beam_size = first_step_mask_sizes[0] // target_prefix_sizes[0]
+        # Step 1
+        masking = pt.zeros((target_prefix_sizes[0], target_prefix_sizes[1] + 1), device=target_prefix.device)
+        masking[:,:target_prefix_sizes[1]] = target_prefix
+        masking.masked_fill_(masking != 0., 1.0)
```

Comment:
I am not quite sure whether to go with clamp function because it is not that intuitive to me how to use ... I googled and found not that easy to figure it out a sequence of min and max formula. Not sure what others think.

pt.where is a good option because it is much easier to understand compared to clamp. However, one nicer thing with masked_fill_ is that it is an in-place operation....

Reply:
clamp just limits the minimum and maximum value, which is what you want here, right? To me the benefit of `clamp` over masking would be that with `clamp` you don't need to explicitly create the full mask of shape `masking.size()`. 

I timed this locally in my laptop and clamp seems to be a bit faster.

If we'd ever want to script/trace that function would inplace still be a benefit?

Reply:
You are right. What I wanted here is simply forcing all non zero ids to 1, which is also equivalent to limiting the minimum and maximum value 0 and 1. I got it now.

I actually also tried to masked_fill_ and and clamp and it is true that clamp is faster on both CPU and GPU (I am very surprised to see this, an interesting learning for me :D). So yeah lets go for clamp indeed. I just changed it. Thanks for the suggestion!


```
>>> import timeit
>>> 
>>> set_up_line = """
... import torch as pt
... pt.manual_seed(1)
... masking = pt.zeros((10, 10))"""
>>> 
>>> executive_line = """
... masking[:,:7] = pt.randint(0, 10, (10, 7))
... masking.masked_fill_(masking != 0., 1.)
... """
>>> 
>>> timeit.timeit(executive_line,setup=set_up_line, number=100000)

2.1035887990146875
>>> 
>>> executive_line = """
... masking[:,:7] = pt.randint(0, 10, (10, 7))
... masking = pt.clamp(masking, 0., 1.)
... """
>>> timeit.timeit(executive_line,setup=set_up_line, number=100000)
1.3465280530508608
```